File: 994-rotting-oranges/994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        n, m, visited,a,b = len(grid), len(grid[0]), set(),0,0
        indices = [(x,y) for x in range(n) for y in range(m)]
        def bfs():
            queue, minutes = deque(), 0
            
            for (x,y) in indices:
                if (x,y) not in visited and grid[x][y] == 2:
                    queue.append((x,y))
                    visited.add((x,y))
                    
            if not queue:
                for (x,y) in indices:
                    if grid[x][y] == 1:
                        return -1
                return 0
            
            while queue:
                n = len(queue)
                for i in range(n):
                    x,y = queue.popleft()
                    grid[x][y] = 2
                    for dirs in DIR:
                        dx, dy = x + dirs[0], y + dirs[1]
                        if (dx,dy) not in visited and inbound(dx,dy) and grid[dx][dy] == 1:
                            queue.append((dx,dy))
                            visited.add((dx,dy))
                minutes += 1
            
            return minutes - 1
        
        
        DIR = [[1,0],[0,1],[0,-1],[-1,0]]
        inbound = lambda x, y: 0 <= x < n and 0 <= y < m
        minutes = bfs()
        
        for (x,y) in indices:
            if (x,y) not in visited and grid[x][y] == 1:
                return -1
            
        return minutes

File: 994-rotting-oranges/994-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int n = grid.size(), m = grid[0].size();
        pair<int,int>dir[] = {{1,0}, {0,1}, {-1, 0}, {0,-1}};
        queue<pair<int,int>>q;
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 2)
                    q.push({i,j});
            }
        }
        
        if (q.empty()) {
            for (int i = 0; i < n; i++)
                for (int j = 0; j < m; j++)
                    if (grid[i][j] == 1)
                        return -1;
            return 0;
        }
        
        auto valid = [&](int x,int y) {
          return (x >= 0 && y >= 0 && x < n && y < m && grid[x][y] == 1);  
        };
        
        int cnt = 0;
        while (!q.empty()) {
            int n = q.size();
            for (int i = 0; i < n; i++) {
                auto u = q.front();
                q.pop();
                grid[u.first][u.second] = 2;
                for (auto [x,y]: dir) {
                    int dx = u.first + x;
                    int dy = u.second + y;
                    if (valid(dx,dy)) {
                        q.push({dx,dy});
                        grid[dx][dy] = 2;
                    }
                }
            }
            cnt++;
        }
        
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                if (grid[i][j] == 1)
                    return -1;
        return cnt - 1;
    }
};